lib/propolis/src/migrate.rs

```diff
@@ -43,28 +43,207 @@ pub enum Migrator<'a> {
+    ///
+    /// The device may be capable of handing differing formats and/or versions
+    /// of said payload, but only one at a time is expected for a given device
+    /// during migration.
+    Single(&'a dyn MigrateSingle),
 
-pub struct MigrateCtx<'a> {
-    pub mem: &'a MemCtx,
+    /// Migration state for the device consists of multiple payloads.  This is
+    /// the case for more complex devices where emulation is composed from
+    /// several abstractions.
+    ///
+    /// One example would be pci-virtio-block, where one payload contains to the
+    /// emulated PCI state, while another contains the virtio state (virtqueues,
+    /// etc).
+    Multi(&'a dyn MigrateMulti),
 }
 
-pub trait Migrate: Send + Sync + 'static {
-    /// Return a serialization of the current device state.
-    fn export(&self, ctx: &MigrateCtx) -> Box<dyn Serialize>;
+/// A device which can be migrated using a single typed payload to represent its
+/// internal state.
+pub trait MigrateSingle: Send + Sync + 'static {
+    fn export(
+        &self,
+        ctx: &MigrateCtx,
+    ) -> Result<PayloadOutput, MigrateStateError>;
+    fn import(
+        &self,
+        offer: PayloadOffer,
+        ctx: &MigrateCtx,
+    ) -> Result<(), MigrateStateError>;
+}
 
-    #[allow(unused_variables)]
-    /// Update the current device state by using the given deserializer.
+/// A device which can be migrated using multiple differently-typed payloads to
+/// represent its internal state.
+pub trait MigrateMulti: Send + Sync {
+    fn export(
+        &self,
+        output: &mut PayloadOutputs,
+        ctx: &MigrateCtx,
+    ) -> Result<(), MigrateStateError>;
     fn import(
         &self,
-        dev: &str,
-        deserializer: &mut dyn Deserializer,
+        offer: &mut PayloadOffers,
         ctx: &MigrateCtx,
+    ) -> Result<(), MigrateStateError>;
+}
+
+/// Additional (borrowed) context data used during device import and export.
+pub struct MigrateCtx<'a> {
+    pub mem: &'a MemCtx,
+}
+
+/// Device state payload (of a given kind/version) offered to the import logic
+/// for that device during a migration.
+pub struct PayloadOffer<'a> {
+    pub kind: &'a str,
+    pub version: u32,
+    pub payload: Box<dyn erased_serde::Deserializer<'a> + 'a>,
+}
+impl<'a> PayloadOffer<'a> {
+    /// Attempt to parse the data in this payload if the offer matches the
+    /// kind/version of a specified Schema
+    pub fn parse<T: Schema<'a>>(&mut self) -> Result<T, MigrateStateError> {
+        if !T::matches(self.kind, self.version) {
+            return Err(MigrateStateError::UnexpectedPayload(
+                self.kind.into(),
+                self.version,
+            ));
+        }
+        let res = erased_serde::deserialize(&mut self.payload)?;
+        Ok(res)
+    }
+}
+
+/// Collection of [`PayloadOffer`] instances, as provided to device state import
+/// logic during a migration.
+pub struct PayloadOffers<'a>(Vec<PayloadOffer<'a>>);
```

Comment:
There's kind of an implicit schema here around the number and kind of payload offers a device will make. Should we make that explicit somehow? For example, suppose Propolis v5 has a `Contraption` device that emits a Foo, Bar, and Baz. Then, in Propolis v6, we simplify or reorganize the `Contraption` so that it emits only a Foo and Bar. Now `Contraption::import` has to make sure to try to `take` a Baz just in case this is an offer from an older Propolis that still includes one (otherwise the import will fail because there will be data remaining).

Is this worth trying to stamp out here? The example above feels very contrived, in the sense that the existing `MigrateMulti` entities' sub-offers are for PCI and virtio state. Because those are guest-visible interfaces, it's hard to see how we could get rid of them version-over-version. I can't think of a concrete example of a shared state type that we might eventually want to refactor away, so I don't want to push for any particular change here, but wanted to at least bring it up for discussion.